src/main/java/org/quiltmc/loader/impl/QuiltLoaderImpl.java

```diff
@@ -172,7 +184,19 @@ public File getConfigDirectory() {
 	}
 
 	public Path getModsDir() {
-		return getGameDir().resolve("mods");
+		if (modsDir == null) {
+			// Should not be null ever
+			refreshModsDir(gameDir);
```

Comment:
This could return a wrong value if someone changes the property of the mods dir via [System.setProperty()](https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#setProperty(java.lang.String,%20java.lang.String)).

We should resolve the default folder, whether it is default location or via system property before handing off control to any mods or mixin config plugins.

Reply:
AFAIK this will only be set once (since nothing sets the `modsDir` field to null) so we don't have to worry about that - after it has been called the first time (like when scanning for mods) it can't change. (Which is also what we want, since only loader plugins should be able to change the mods directory in the future, and definitely not at runtime)

Reply:
So Alex, the value is set lazily. When someone first queries the mods directory, this is called. So the first person to set the system property and call `getModsDir` can change the return value to be incorrect. Easily done via pre launch if no one else is before you or in a mixin plugin.

Reply:
The `getModsDir()` method is called to determine what mods to load - I.E. in order to determine what mixin plugins might be available, or what pre-launch entrypoints need to be called. So it's set lazily, yes - however it's set *before* either of those are even loaded.

Reply:
Okay, what about the config dir then? It seems to suffer a similar impl case but without the early loading that mod resolution does.

Reply:
Currently `DependencyOverrides` calls it, also before mods are fully loaded.

Reply:
Hmm, okay so all cases are covered.

Reply:
> So Alex, the value is set lazily. When someone first queries the mods directory, this is called. So the first person to set the system property and call `getModsDir` can change the return value to be incorrect. Easily done via pre launch if no one else is before you or in a mixin plugin.

Nope, that's a fail safe in case something cursed happens since the mod directory is set and initialized right as the loader impl instance is setup, way before mod loading